flamingpy/noise/cv.py

```diff
@@ -0,0 +1,649 @@
+        the initial (ideal) measurement outcomes.
+
+        This method modifies self.egraph.
+        """
+        macro_graph = self.egraph
+        perfect_points = macro_graph.graph.get("perfect_points")
+        if perfect_points:
+            perfect_inds = [macro_graph.to_indices[point] for point in perfect_points]
+        else:
+            perfect_inds = None
+        noise_model["perfect_inds"] = perfect_inds
+        noise_model["sampling_order"] = "two-step"
+        self.apply_noise(noise_model)
+
+    def _permute_indices_and_label(self):
+        """Obtain permuted indices and set type of reduced node.
+
+        For each macronode, permute indices so that the first encountered GKP state
+        comes first, designating it as the 'star' ('central') mode. The first
+        index in the resulting list and every four indices thereafter correspond
+        to star modes. The rest are 'planets' ('satellite' modes).
+
+        For each star, associate a 'body_index' of 1, and 2, 3, and 4 for the
+        subsequent planets. Additionally, if a macronode contains at least one GKP
+        state, label the reduced state as 'GKP' (otherwise 'p').
+
+        This method sets the attribute self.permuted_inds to the permuted
+        indices and modifies self.egraph and self.reduced_graph.
+        """
+        N = self._N
+        macro_graph = self.egraph
+        to_points = macro_graph.to_points
+        # A list of permuted indices where each block of four
+        # corresponds to [star, planet, planet, planet].
+        permuted_inds = np.empty(N, dtype=np.int32)
+        for i in range(0, N - 3, 4):
+            # Indices of GKP micronodes in macronode i.
+            gkps = []
+            for j in range(4):
+                micronode = to_points[i + j]
+                if macro_graph.nodes[micronode]["state"] == "GKP":
+                    gkps.append(j)
+            centre_point = tuple(round(i) for i in micronode)
+            if gkps:
+                star_ind, reduced_state = i + gkps[0], "GKP"
+            else:
+                star_ind, reduced_state = i, "p"
+            # Set type of node in the reduced lattice as a p-squeezed
+            # state if all micronodes are p, else GKP.
+            self.reduced_graph.nodes[centre_point]["state"] = reduced_state
+            # Old and permuted indices of all micronodes in macronode i.
+            old_inds = [i, i + 1, i + 2, i + 3]
+            old_inds.pop(star_ind - i)
+            new_inds = [star_ind] + old_inds
+            # Associate a 'body index' (1 to 4) to each micronode,
+            # with 1 being the star index and the rest being planets.
+            k = 1
+            for ind in new_inds:
+                macro_graph.nodes[to_points[ind]]["body_index"] = k
+                k += 1
+            permuted_inds[[i, i + 1, i + 2, i + 3]] = new_inds
+        self.permuted_inds = permuted_inds
```

Comment:
Regarding this function, I think it can be written as 

```python
def _permute_indices(self):
    permuted_inds = np.empty(self._N, dtype=np.int32)
    inds = np.reshape(np.arange(self._N), (N // 4, 4))
    self.permuted_inds = np.flatten(np.apply_along_axis(self._permute_macronode_indices, 1, inds))
    
def _permute_macronode_indices(self, indices):
    gkps = [self.egraph.nodes[self.to_points[ind]]["state"] == "GKP" for ind in indices]
    return np.concat(indices[gkps], indices[~gkps])

def _apply_labels(self):
    for ind in self.permuted_inds:
        self.egraph.nodes[self.to_points[ind]]["body_index"] = ind % 4
    for ind in self.permuted_inds[::4]:
        point = self.to_points[ind]
        centre_point = tuple(round(i) for i in point)
        self.reduced_graph.nodes[centre_point]["state"] = self.egraph.nodes[point]["state"]
```

This version is smaller and should be faster since it use numpy functions more.

Reply:
With some modifications (array.flatten() instead of np.flatten, np.concatenate((indices[gkps], indices[not ind for ind in gkps])), doesn't seem to produce the right result. It's promising but I would have to do a slightly deeper dive.

Reply:
I made the assumption that the ordering of gkps nodes is not important as long as the first state is a gkp one if it exists. Maybe this is not a fair assumption.

Reply:
It's a fair assumption! The mistake was in this line:
```
 for ind in self.permuted_inds:
        self.egraph.nodes[self.to_points[ind]]["body_index"] = ind % 4
```
Which I fixed to:

```
        for i, ind in enumerate(self.permuted_inds):
            self.egraph.nodes[self._to_points[ind]]["body_index"] = i % 4 + 1
```